### ISSUE TITLE
Misleading error: cannot borrow captured outer variable in an `FnMut` closure as mutable

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#52490]
[This code](https://play.rust-lang.org/?gist=682ebd6c22cceb84af45f40c2c25d8d9&version=stable&mode=debug&edition=2015)
```rust
fn foo(_f: impl FnMut()) {}

fn main() {
    let v = vec![0];
    foo(move || { let _x = &mut v[..]; });
}
```
fails to compile saying
```
error[E0596]: cannot borrow captured outer variable in an `FnMut` closure as mutable
 --> src/main.rs:5:33
  |
5 |     foo(move || { let _x = &mut v[..]; });
  |                                 ^
```
It took me a while to understand what's going on: The actual problem is not the `FnMut`, and it's not the place the error points at, it's that the outer `v` should be `let mut v`.

The way the error is phrased, it sounds like the `FnMut` is the problem here, so I was really confused.

**Comment by kulshrax** [comment]
I just ran into this same issue. Based on the error message, I ended up using an unnecessary `Mutex` until I ran `rustc --explain E0596` and realized the issue was the mutability of the outer variable. I think a simple fix here would just be to add the word "immutable" to the error message:

> cannot borrow captured immutable outer variable in an `FnMut` closure as mutable

**Comment by estebank** [comment]
Current error is more illuminating:

```
error[E0596]: cannot borrow `v` as mutable, as it is not declared as mutable
 --> src/main.rs:5:33
  |
4 |     let v = vec![0];
  |         - help: consider changing this to be mutable: `mut v`
5 |     foo(move || { let _x = &mut v[..]; });
  |                                 ^ cannot borrow as mutable
```

**Comment by RalfJung** [comment]
Yeah that looks awesome. :)

I hope there's a test?
